[RESTEASY-2312] Introduce support for Spring REST annotation handling

geoand opened the pull request:
cc @asoldano @patriot1burke

patriot1burke commented on resteasy-spring-web/src/main/java/org/jboss/resteasy/web/ResponseEntityReturnTypeHandler.java:
```diff
@@ -0,0 +1,46 @@
+package org.jboss.resteasy.web;
```
package should have spring in it.  org.jboss.resteasy.spring.web or something

patriot1burke commented on resteasy-spring-web/src/main/java/org/jboss/resteasy/web/ResponseStatusBuiltResponseProcessor.java:
```diff
@@ -0,0 +1,29 @@
+
+import javax.ws.rs.core.Response;
+
+public class ResponseStatusBuiltResponseProcessor implements BuiltResponseProcessor {
```
I think this could be implemented as a ContainerResponseFilter instead.

patriot1burke commented:
I didn't see any documentation either.

geoand commented:
Thanks for the comments @patriotburke1! I'll address them tomorrow 😊

patriot1burke commented on resteasy-core-spi/src/main/java/org/jboss/resteasy/spi/BuiltResponseProcessor.java:
```diff
@@ -0,0 +1,13 @@
+
+import javax.ws.rs.core.Response;
+
+public interface BuiltResponseProcessor {
```
Same here, I think this can be implemented as a combination of DynamicFeature and ContainerFilterResponse.  You scan annotatoins and view return types within DynamicFeature.  You add a ContainerFilterResponse for the method if its a Spring one.  This filter handles the transformation between the spring return object and jaxrs Response.

geoand replied:
Thanks for the tip, looking into it

geoand commented:
@patriot1burke I updated the PR to get rid of the 2 SPI's I had added that could replaced with the `DynamicFeature` and it worked like I charm! Thanks for the great tip!

I also addressed the wrong package issue you mentioned.

As for the documentation, where would you like me to add it and what do you think should be added? 
The thing with the feature as I see it is that it's only really meant to be used by Quarkus. Of course it could be used outside of it, but there are a few things that would be need to be configured manually. Do you think such configuration should be included?

Thanks

geoand commented:
@rsearls Thanks

I have noticed that too and it's due to the different line-ending characters. I tried to change it but all my attempts failed. 
I will give it another go however and see if I can find something. You wouldn't happen to have any tips would you?

geoand commented:
It seems like my IDE is doing something weird with the line endings, I'll see how to fix it and then update the PR

geoand commented:
@rsearls formatting issues should now be fixed and the PR should be much easier to review

patriot1burke commented:
What's manual?

If you add META-INF/services/javax.ws.rs.ext.Providers with the FQN if your DynamicFeature, it should just work.  Also, you could create a regular Feature to register the param injectors too.

geoand commented:
@patriot1burke the way I have things, "telling" RESTEasy which classes are Spring Controllers is manual (by setting `resteasy.scanned.resource.classes.with.builder`).
The other parts of course can be registered as you mention.

geoand commented:
The test failure seems unrelated:

```
The command "travis_wait 60 mvn -B -Ptravis,jboss-repository -fae -Dserver.version=$SERVER_VERSION ${ELYTRON:+-Delytron} install" exited with 1.
```

geoand commented:
I'd be happy to add whatever docs you think are relevant for this use case

geoand commented:
@asoldano thanks!

I created a JIRA [here](https://issues.jboss.org/browse/RESTEASY-2312). Should I add any more information to it? I will also update the PR to be against master instead of `4.1`.
Futhermore I will add basic information in docs. Any preference where that should be?
 
As for the tests, I believe they should be separate, since this feature is different from the other Spring stuff in that it doesn't actually involve Spring, only reading Spring metadata. WDYT?

geoand commented:
@asoldano PR now is against `master`. I also squashed the commits into a single one

geoand commented:
I also added some basic documentation in the Spring Integration section

asoldano commented:
OK, thanks @geoand .
Last question I have is about Spring version. I'm locally testing an upgrade of the whole project from 4.3.22.RELEASE to the 5.0.14.RELEASE version you're pulling here... however is that choose for a specific reason? Would 5.1.x be fine as well?

geoand commented:
@asoldano no particular reason for the version I chose. It should work with anything higher than 4.3.x.
5.1.x should be perfectly fine as well

asoldano commented:
@geoand ok, thanks

geoand commented:
I closed and reopened the PR a few times to get restart Travis - the failures seem totally irrelevant

liweinan commented:
great!
